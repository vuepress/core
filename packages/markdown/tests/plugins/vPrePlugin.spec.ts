import MarkdownIt from 'markdown-it'
import { describe, expect, it, vi } from 'vitest'
import { vPrePlugin } from '../../src/index.js'

const CODE_FENCE = '```'

describe('plugin options', () => {
  const source = `\
${CODE_FENCE}js
const a = 1
${CODE_FENCE}

\`inline\`
`
  it('should process code with default options', () => {
    const md = MarkdownIt().use(vPrePlugin)

    expect(md.render(source)).toMatchSnapshot()
  })

  it('should disable `block`', () => {
    const md = MarkdownIt().use(vPrePlugin, { block: false })

    expect(md.render(source)).toMatchSnapshot()
  })

  it('should disable `inline`', () => {
    const md = MarkdownIt().use(vPrePlugin, { inline: false })

    expect(md.render(source)).toMatchSnapshot()
  })

  it('should disable `block` and `inline`', () => {
    const md = MarkdownIt().use(vPrePlugin, { block: false, inline: false })

    expect(md.render(source)).toMatchSnapshot()
  })
})

describe(':v-pre / :no-v-pre', () => {
  const source = `\
${CODE_FENCE}js:v-pre
const a = 1
${CODE_FENCE}

${CODE_FENCE}js:no-v-pre
const a = 1
${CODE_FENCE}

${CODE_FENCE}js{1,2}:v-pre
const a = 1
${CODE_FENCE}

${CODE_FENCE}js{1,2}:no-v-pre
const a = 1
${CODE_FENCE}

${CODE_FENCE}js:other-syntax:v-pre
const a = 1
${CODE_FENCE}

${CODE_FENCE}js:other-syntax:no-v-pre
const a = 1
${CODE_FENCE}

${CODE_FENCE}js
const a = 1
${CODE_FENCE}
`
  it('should work if `block` is enabled by default', () => {
    const md = MarkdownIt().use(vPrePlugin)

    expect(md.render(source)).toMatchSnapshot()
  })

  it('should work if `block` is disabled', () => {
    const md = MarkdownIt().use(vPrePlugin, { block: false })

    expect(md.render(source)).toMatchSnapshot()
  })
})

describe('syntax highlighting', () => {
  const source = `\
${CODE_FENCE}js:v-pre
const a = 1
${CODE_FENCE}

${CODE_FENCE}js:no-v-pre
const a = 1
${CODE_FENCE}

${CODE_FENCE}js{1,2}:v-pre
const a = 1
${CODE_FENCE}

${CODE_FENCE}js{1,2}:no-v-pre
const a = 1
${CODE_FENCE}

${CODE_FENCE}js:other-syntax:v-pre
const a = 1
${CODE_FENCE}

${CODE_FENCE}js:other-syntax:no-v-pre
const a = 1
${CODE_FENCE}

${CODE_FENCE}js
const a = 1
${CODE_FENCE}
`
  it('should work highlighted code is wrapped with `<pre>`', () => {
    const highlight = vi.fn(
      (code, lang) =>
        `<pre><code>highlighted code: ${code}, lang: ${lang}</code></pre>`,
    )

    const md = MarkdownIt({ highlight }).use(vPrePlugin)

    expect(md.render(source)).toMatchSnapshot()
    expect(highlight).toHaveBeenCalledTimes(7)
  })

  it('should work if highlighted code is not wrapped with `<pre>`', () => {
    const highlight = vi.fn(
      (code, lang) => `highlighted code: ${code}, lang: ${lang}`,
    )
    const md = MarkdownIt({ highlight }).use(vPrePlugin)

    expect(md.render(source)).toMatchSnapshot()
    expect(highlight).toHaveBeenCalledTimes(7)
  })
})
