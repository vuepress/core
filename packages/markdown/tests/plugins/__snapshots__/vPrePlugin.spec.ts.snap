// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`:v-pre / :no-v-pre > should work if \`block\` is disabled 1`] = `
"<pre v-pre><code class="language-js:v-pre">const a = 1
</code></pre>
<pre><code class="language-js:no-v-pre">const a = 1
</code></pre>
<pre v-pre><code class="language-js{1,2}:v-pre">const a = 1
</code></pre>
<pre><code class="language-js{1,2}:no-v-pre">const a = 1
</code></pre>
<pre v-pre><code class="language-js:other-syntax:v-pre">const a = 1
</code></pre>
<pre><code class="language-js:other-syntax:no-v-pre">const a = 1
</code></pre>
<pre><code class="language-js">const a = 1
</code></pre>
"
`;

exports[`:v-pre / :no-v-pre > should work if \`block\` is enabled by default 1`] = `
"<pre v-pre><code class="language-js:v-pre">const a = 1
</code></pre>
<pre><code class="language-js:no-v-pre">const a = 1
</code></pre>
<pre v-pre><code class="language-js{1,2}:v-pre">const a = 1
</code></pre>
<pre><code class="language-js{1,2}:no-v-pre">const a = 1
</code></pre>
<pre v-pre><code class="language-js:other-syntax:v-pre">const a = 1
</code></pre>
<pre><code class="language-js:other-syntax:no-v-pre">const a = 1
</code></pre>
<pre v-pre><code class="language-js">const a = 1
</code></pre>
"
`;

exports[`plugin options > should disable \`block\` 1`] = `
"<pre><code class="language-js">const a = 1
</code></pre>
<p><code v-pre>inline</code></p>
"
`;

exports[`plugin options > should disable \`block\` and \`inline\` 1`] = `
"<pre><code class="language-js">const a = 1
</code></pre>
<p><code>inline</code></p>
"
`;

exports[`plugin options > should disable \`inline\` 1`] = `
"<pre v-pre><code class="language-js">const a = 1
</code></pre>
<p><code>inline</code></p>
"
`;

exports[`plugin options > should process code with default options 1`] = `
"<pre v-pre><code class="language-js">const a = 1
</code></pre>
<p><code v-pre>inline</code></p>
"
`;

exports[`syntax highlighting > should work highlighted code is wrapped with \`<pre>\` 1`] = `
"<pre v-pre><code>highlighted code: const a = 1
, lang: js:v-pre</code></pre>
<pre><code>highlighted code: const a = 1
, lang: js:no-v-pre</code></pre>
<pre v-pre><code>highlighted code: const a = 1
, lang: js{1,2}:v-pre</code></pre>
<pre><code>highlighted code: const a = 1
, lang: js{1,2}:no-v-pre</code></pre>
<pre v-pre><code>highlighted code: const a = 1
, lang: js:other-syntax:v-pre</code></pre>
<pre><code>highlighted code: const a = 1
, lang: js:other-syntax:no-v-pre</code></pre>
<pre v-pre><code>highlighted code: const a = 1
, lang: js</code></pre>
"
`;

exports[`syntax highlighting > should work if highlighted code is not wrapped with \`<pre>\` 1`] = `
"<pre v-pre><code class="language-js:v-pre">highlighted code: const a = 1
, lang: js:v-pre</code></pre>
<pre><code class="language-js:no-v-pre">highlighted code: const a = 1
, lang: js:no-v-pre</code></pre>
<pre v-pre><code class="language-js{1,2}:v-pre">highlighted code: const a = 1
, lang: js{1,2}:v-pre</code></pre>
<pre><code class="language-js{1,2}:no-v-pre">highlighted code: const a = 1
, lang: js{1,2}:no-v-pre</code></pre>
<pre v-pre><code class="language-js:other-syntax:v-pre">highlighted code: const a = 1
, lang: js:other-syntax:v-pre</code></pre>
<pre><code class="language-js:other-syntax:no-v-pre">highlighted code: const a = 1
, lang: js:other-syntax:no-v-pre</code></pre>
<pre v-pre><code class="language-js">highlighted code: const a = 1
, lang: js</code></pre>
"
`;
